### DISCUSSION TITLE
AutoNat

### DISCUSSION THREAD

**Discussion by kehiy** [libp2p/rust-libp2p#4500] [Q&A] [answered]
Hi,
About AutoNat, when a private node tries to determine whether it's public or private.
When it sends a packet to a public node when the public node tries to dial it and send the result to the private node the private node NAT also has a `5-tuple` and it should work(?).
And if it doesn't, how they send the result to a private node?

**Comment by thomaseizinger** [comment, marked as answer]
Two things: Our current AutoNAT implementation is (unfortunately by design) a bit flawed because we might accidentally be hole-punching and thus falsely reporting a "no NAT" status. To fix this, there is a V2 currently in the works: https://github.com/libp2p/specs/pull/538

> And if it doesn't, how they send the result to a private node?

The node that wants its NAT status figured out dialed the AutoNAT server. This means the server can still open a stream back to the node and report the result of the dial attempt. See the specs for details: https://github.com/libp2p/specs/tree/master/autonat

**Reply by kehiy** [reply]
So, if it can open a stream back
How does it check whether the node that sends the request is dialable or not?

**Reply by thomaseizinger** [reply]
The spec explains that in detail! :)

**Reply by kehiy** [reply]
thanksss
